File: Sources/CountriesFlagsObjC/source/Countries/UIImage+UnitedStatesOfAmerica.m
//
//  UIImage+UnitedStatesOfAmerica.m
//  CountriesFlags
//

#import "Countries/UIImage+UnitedStatesOfAmerica.h"

#import "Extensions/UIBezierPath+Star.h"
#import "Extensions/CGSize+Extension.h"

@implementation UIImage (UnitedStatesOfAmerica)

+ (UIImage * _Nullable)flagUnitedStatesOfAmericaWithSize:(CGSize)size {
    UIColor* colorBlue = [UIColor colorWithRed:10.0f/255.0f
                                         green:49.0f/255.0f
                                          blue:97.0f/255.0f
                                         alpha:1.0f];
    UIColor* colorRed = [UIColor colorWithRed:179.0f/255.0f
                                        green:25.0f/255.0f
                                         blue:66.0f/255.0f
                                        alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    
    UIGraphicsBeginImageContextWithOptions(size, NO, 0.0f);
    
    [colorRed setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorWhite setFill];
    for (int index = 1; index <= 13; index += 2) {
        UIRectFill(CGRectMake(0.0f, size.height / 13.0f * index, size.width, size.height / 13.0f));
    }
    
    [colorBlue setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width / 5.0f * 2.0f, size.height / 13.0f * 7.0f));

    [colorWhite setFill];
    CGFloat xOffset = size.width / 5.0f * 2.0f / 12.0f;
    CGFloat yOffset = size.height / 13.0f * 7.0f / 10.0f;
    CGFloat starRadius = size.height * 0.0616f / 2.0f;
    for (int indexY = 1; indexY <= 5; indexY++) {
        for (int indexX = 0; indexX <= 5; indexX++) {
            CGSize size = CGSizeMake(xOffset * 2.0f + xOffset * (indexX * 2.0f) * 2.0f, yOffset * indexY * 4.0f - yOffset * 2.0f);
            [[UIBezierPath starAtPoint:CGSizeMidPoint(size)
                                radius:starRadius
                                  type:FlagStarTypeFivePointed] fill];
        }
    }
    for (CGFloat indexY = 1.5f; indexY <= 5; indexY++) {
        for (CGFloat indexX = 0.5; indexX <= 5; indexX++) {
            CGSize size = CGSizeMake(xOffset * 2.0f + xOffset * (indexX * 2.0f) * 2.0f, yOffset * indexY * 4.0f - yOffset * 2.0f);
            [[UIBezierPath starAtPoint:CGSizeMidPoint(size)
                                radius:starRadius
                                  type:FlagStarTypeFivePointed] fill];
        }
    }
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+UnitedStatesOfAmerica.swift
//
//  UIImage+UnitedStatesOfAmerica.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_the_United_States"wikipedia"
     
     Generates the flag of the United States.
     For more information see [wikipedia].

     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of the United States.
     */
    internal static func flagUnitedStatesOfAmerica(with size: CGSize) -> UIImage? {
        // Federal Specification DDD-F-416F cloth color specifications
        let colorBlue = UIColor(red: 10.0 / 255.0,
                                green: 49.0 / 255.0,
                                blue: 97.0 / 255.0,
                                alpha: 1.0)
        let colorRed = UIColor(red: 179.0 / 255.0,
                               green: 25.0 / 255.0,
                               blue: 66.0 / 255.0,
                               alpha: 1.0)
        let colorWhite = UIColor.white

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorRed.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorWhite.setFill()
        for index in stride(from: 1.0, through: 13.0, by: 2.0) {
            UIRectFill(CGRect(origin: CGPoint(x: 0.0, y: size.height / 13.0 * index),
                              size: CGSize(width: size.width, height: size.height / 13.0)))
        }

        colorBlue.setFill()
        UIRectFill(CGRect(origin: .zero,
                          size: CGSize(width: size.width / 5.0 * 2.0, height: size.height / 13.0 * 7.0)))

        colorWhite.setFill()
        let xOffset = size.width / 5.0 * 2.0 / 12.0
        let yOffset = size.height / 13.0 * 7.0 / 10.0
        let starRadius = size.height * 0.0616 / 2.0
        for indexY in stride(from: 1.0, through: 5.0, by: 1.0) {
            for indexX in stride(from: 0.0, through: 5.0, by: 1.0) {
                let size = CGSize(width: xOffset * 2.0 + xOffset * (indexX * 2.0) * 2.0,
                                  height: yOffset * indexY * 4.0 - yOffset * 2.0)
                UIBezierPath.star(at: size.midPoint, radius: starRadius).fill()
            }
        }
        for indexY in stride(from: 1.5, through: 5.0, by: 1.0) {
            for indexX in stride(from: 0.5, through: 5.0, by: 1.0) {
                let size = CGSize(width: xOffset * 2.0 + xOffset * (indexX * 2.0) * 2.0,
                                  height: yOffset * indexY * 4.0 - yOffset * 2.0)
                UIBezierPath.star(at: size.midPoint, radius: starRadius).fill()
            }
        }

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
